pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -943,11 +960,126 @@ impl Drop for ImageLayerWriter {
     }
 }
 
+pub struct ImageLayerIterator<'a, 'ctx> {
+    image_layer: &'a ImageLayerInner,
+    ctx: &'ctx RequestContext,
+    next_batch_start_key: Key,
+    key_values_batch: Vec<(Key, Lsn, Value)>,
+    next_idx_in_batch: usize,
+    is_end: bool,
+    /// Limit of number of keys per batch
+    batch_size: usize,
+    /// Limit of read size per batch
+    max_read_size: u64,
+}
+
+impl<'a, 'ctx> ImageLayerIterator<'a, 'ctx> {
+    /// Retrieve a batch of key-value pairs into the iterator buffer.
+    async fn next_batch(&mut self) -> anyhow::Result<()> {
+        assert!(self.next_idx_in_batch >= self.key_values_batch.len());
+        assert!(!self.is_end);
+        self.key_values_batch.clear();
+        self.next_idx_in_batch = 0;
+
+        let block_reader = FileBlockReader::new(&self.image_layer.file, self.image_layer.file_id);
+        let tree_reader = DiskBtreeReader::new(
+            self.image_layer.index_start_blk,
+            self.image_layer.index_root_blk,
+            &block_reader,
+        );
+        let mut search_key = [0; KEY_SIZE];
+        self.next_batch_start_key
+            .write_to_byte_slice(&mut search_key);
+        // We want to have exactly one read syscall (plus several others for index lookup) for each `next_batch` call.
+        // Therefore, we enforce `self.max_read_size` by ourselves instead of using the VectoredReadPlanner's capability,
+        // to avoid splitting into two I/Os.
+        let mut read_planner = VectoredReadPlanner::new_without_max_limit();
+        let mut cnt = 0;
+        let mut start_pos = None;
+        let mut range_end_handled = false;
+        // TODO: dedup with vectored read?
+        tree_reader
+            .visit(
+                &search_key,
+                VisitDirection::Forwards,
+                |raw_key, offset| {
```

Comment:
@VladLazar when you introduced vectored read, I remember I was suggesting some abstraction wrt the visitor during reviews.
Unfortunaly I don't remember what exactly. Do you? 

And generally, do you have any thoughts on whether we can extract some common code?

Reply:
Not really, but I remember the idea of a stream being floated. That was implemented: `DiskBtreeReader::get_stream_from`. If possible, I'd like new code to use the streaming interface since the legacy read path is the only user of `DiskBtreeReader::visit` and I was planning on removing the function.